// @atlassian/aui/lib/js/aui/banner.js
(function (global, factory) {
    if (typeof define === 'function' && define.amd) {
        define(['exports', 'module', './jquery', './internal/animation', './internal/amdify', './internal/globalize', './template'], factory);
    } else if (typeof exports !== 'undefined' && typeof module !== 'undefined') {
        factory(exports, module, require('./jquery'), require('./internal/animation'), require('./internal/amdify'), require('./internal/globalize'), require('./template'));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, mod, global.$, global.animation, global.amdify, global.globalize, global.template);
        global.banner = mod.exports;
    }
})(this, function (exports, module, _jquery, _internalAnimation, _internalAmdify, _internalGlobalize, _template) {
    'use strict';

    function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

    var _$ = _interopRequireDefault(_jquery);

    var _amdify = _interopRequireDefault(_internalAmdify);

    var _globalize = _interopRequireDefault(_internalGlobalize);

    var _template2 = _interopRequireDefault(_template);

    var ID_BANNER_CONTAINER = 'header';

    function banner(options) {
        var $banner = renderBannerElement(options);

        pruneBannerContainer();
        insertBanner($banner);

        return $banner[0];
    }

    function renderBannerElement(options) {
        var html = '<div class="aui-banner aui-banner-{type}" role="banner">' + '{body}' + '</div>';

        var $banner = (0, _$['default'])((0, _template2['default'])(html).fill({
            'type': 'error',
            'body:html': options.body || ''
        }).toString());

        return $banner;
    }

    function pruneBannerContainer() {
        var $container = findContainer();
        var $allBanners = $container.find('.aui-banner');

        $allBanners.get().forEach(function (banner) {
            var isBannerAriaHidden = banner.getAttribute('aria-hidden') === 'true';
            if (isBannerAriaHidden) {
                (0, _$['default'])(banner).remove();
            }
        });
    }

    function findContainer() {
        return (0, _$['default'])('#' + ID_BANNER_CONTAINER);
    }

    function insertBanner($banner) {
        var $bannerContainer = findContainer();
        if (!$bannerContainer.length) {
            throw new Error('You must implement the application header');
        }

        $banner.prependTo($bannerContainer);
        (0, _internalAnimation.recomputeStyle)($banner);
        $banner.attr('aria-hidden', 'false');
    }

    (0, _amdify['default'])('aui/banner', banner);
    (0, _globalize['default'])('banner', banner);
    module.exports = banner;
});
