// @atlassian/aui/lib/js/aui/tooltip.js
(function (global, factory) {
    if (typeof define === 'function' && define.amd) {
        define(['exports', './jquery', '../../js-vendor/jquery/jquery.tipsy'], factory);
    } else if (typeof exports !== 'undefined') {
        factory(exports, require('./jquery'), require('../../js-vendor/jquery/jquery.tipsy'));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, global.$, global.jquery);
        global.tooltip = mod.exports;
    }
})(this, function (exports, _jquery, _jsVendorJqueryJqueryTipsy) {
    'use strict';

    function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

    var _$ = _interopRequireDefault(_jquery);

    _$['default'].fn.tooltip = function (options) {
        return this.each(function () {
            var $this = (0, _$['default'])(this);
            var allOptions = _$['default'].extend({}, _$['default'].fn.tooltip.defaults, options);

            // Pass string values straight to tipsy
            if (typeof options === 'string') {
                $this.tipsy(options);

                if (options === 'destroy') {
                    if (allOptions.live) {
                        (0, _$['default'])($this.context).off('.tipsy', $this.selector);
                    } else {
                        $this.unbind('.tipsy');
                    }
                }
                return;
            }

            $this.tipsy(allOptions);

            if (allOptions.hideOnClick && (allOptions.trigger === 'hover' || !allOptions.trigger && $this.tipsy.defaults.trigger === 'hover')) {
                var onClick = function onClick() {
                    $this.tipsy('hide');
                };

                if (allOptions.live) {
                    (0, _$['default'])($this.context).on('click.tipsy', this.selector, onClick);
                } else {
                    $this.bind('click.tipsy', onClick);
                }
            }
        });
    };

    _$['default'].fn.tooltip.defaults = {
        opacity: 1.0,
        offset: 1,
        delayIn: 500,
        hoverable: true,
        hideOnClick: true,
        aria: true
    };
});
